tiatoolbox/tools/stainextract.py

```diff
@@ -14,35 +14,207 @@
 # along with this program; if not, write to the Free Software Foundation,
 # Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 #
+# This file contains code inspired by StainTools
+# [https://github.com/Peter554/StainTools] written by Peter Byfield.
+#
 # The Original Code is Copyright (C) 2020, TIALab, University of Warwick
 # All rights reserved.
 # ***** END GPL LICENSE BLOCK *****
 
-"""Stain normalisation utilities used in tiatoolbox"""
+"""Stain normalisation stain matrix extraction."""
 import numpy as np
+from sklearn.decomposition import DictionaryLearning
+
+from tiatoolbox.utils.transforms import convert_RGB2OD
+from tiatoolbox.utils.misc import get_luminosity_tissue_mask
 
 
 class CustomExtractor:
-    """Get the user-defined stain matrix"""
+    """Get the user-defined stain matrix.
+
+    This class contains code inspired by StainTools
+    [https://github.com/Peter554/StainTools] written by Peter Byfield.
+
+    Examples:
+        >>> from tiatoolbox.tools.staiextract import CustomExtractor()
+        >>> extractor = CustomExtractor(stain_matrix)
+        >>> stain_matrix = extractor.get_stain_matrix(img)
+
+    """
 
     def __init__(self, stain_matrix):
         self.stain_matrix = stain_matrix
         if self.stain_matrix.shape != (2, 3) and self.stain_matrix.shape != (3, 3):
             raise Exception("Stain matrix must be either (2,3) or (3,3)")
 
     def get_stain_matrix(self, _):
+        """Get the user defined stain matrix.
+
+        Returns:
+            ndarray: user defined stain matrix.
+
+        """
         return self.stain_matrix
 
 
 class RuifrokExtractor:
     """Get the stain matrix as defined in:
```

Comment:
The first line of docstring should end with a period other wise deepsource does not like it. You can rewrite it as

Ruifork stain extractor.

Get the stain matrix as defined in: ...

Reply:
Fixed